src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ImmutableArrayExtensions.cs

```diff
@@ -2,13 +2,204 @@
+            if (count < 0 || startIndex - count + 1 < 0)
+            {
+                ThrowHelper.ThrowCountArgumentOutOfRange_ArgumentOutOfRange_Count();
+            }
+
+            var endIndex = startIndex - count;
+            for (var i = startIndex; i > endIndex; i--)
+            {
+                if (match(items[i]))
+                {
+                    return i;
+                }
+            }
+
+            return -1;
+        }
+
+        public static int FindLastIndex<T, TArg>(this ImmutableArray<T> items, Func<T, TArg, bool> match, TArg arg)
+            => FindLastIndex(items, items.Length - 1, items.Length, match, arg);
+
+        public static int FindLastIndex<T, TArg>(this ImmutableArray<T> items, int startIndex, Func<T, TArg, bool> match, TArg arg)
+            => FindLastIndex(items, startIndex, startIndex + 1, match, arg);
+
+        /// <seealso cref="SegmentedList{T}.FindLastIndex(int, int, Predicate{T})"/>
+        public static int FindLastIndex<T, TArg>(this ImmutableArray<T> items, int startIndex, int count, Func<T, TArg, bool> match, TArg arg)
+        {
+            if (match == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.match);
+            }
+
+            if (items.Length == 0)
+            {
+                // Special case for 0 length ImmutableArray
+                if (startIndex != -1)
+                {
+                    ThrowHelper.ThrowStartIndexArgumentOutOfRange_ArgumentOutOfRange_IndexMustBeLess();
+                }
+            }
+            else
+            {
+                // Make sure we're not out of range
+                if ((uint)startIndex >= (uint)items.Length)
+                {
+                    ThrowHelper.ThrowStartIndexArgumentOutOfRange_ArgumentOutOfRange_IndexMustBeLess();
+                }
+            }
+
+            // 2nd half of this also catches when startIndex == MAXINT, so MAXINT - 0 + 1 == -1, which is < 0.
+            if (count < 0 || startIndex - count + 1 < 0)
+            {
+                ThrowHelper.ThrowCountArgumentOutOfRange_ArgumentOutOfRange_Count();
+            }
+
+            var endIndex = startIndex - count;
+            for (var i = startIndex; i > endIndex; i--)
+            {
+                if (match(items[i], arg))
+                {
+                    return i;
+                }
+            }
+
+            return -1;
+        }
+
+        public static Range FindRange<T>(this ImmutableArray<T> items, Predicate<T> match)
+        {
+            var firstIndex = FindIndex(items, match);
+            if (firstIndex < 0)
+                return default;
+
+            var lastIndex = FindLastIndex(items, match);
+            Contract.ThrowIfFalse(lastIndex >= firstIndex);
+            return new Range(firstIndex, lastIndex + 1);
+        }
+
+        public static Range FindRange<T, TArg>(this ImmutableArray<T> items, Func<T, TArg, bool> match, TArg arg)
```

Comment:
Is this generally useful? Seems like very specialized helper as it assumes that the matching items are in a single contiguous block.

Reply:
I'm not sure. It isn't necessarily for cases where all items in the range match the function; I should document this so it's clear that it's just a combination of `FindFirstIndex` and `FindLastIndex` which is returned as a `Range`.

Reply:
➡️ Switched to different approach